File: src/Befunge98Strategy.java
/**
 * @file Befunge98Strategy.java
 * Strategy for Befunge-98 commands.
 * @author Conlan Wesson
 */

import java.time.LocalDateTime;

class Befunge98Strategy extends Unefunge98Strategy{
	public Befunge98Strategy(Field f, InstructionPointer i, StackStack s){
		super(f, i, s);
	}
	
	public boolean execute(char cmd){
		switch(cmd){
			// Befunge-93
			case '^':
				ip.setDir(0, -1); break;
			case 'v':
				ip.setDir(0, 1); break;
			case '|':
				if(stack.top().pop() == 0){
					ip.setDir(0, 1);
				}else{
					ip.setDir(0, -1);
				}
				break;
			case '?':{
				int r = rnd.nextInt(4);
				switch(r){
					case 0:
						ip.setDir(0, 1); break;
					case 1:
						ip.setDir(-1, 0); break;
					case 2:
						ip.setDir(1, 0); break;
					case 3:
						ip.setDir(0, -1); break;
				}
				break;
			}
			
			case '[':
				ip.left(); break;
			case ']':
				ip.right(); break;
			case 'w':{
				int b = stack.top().pop();
				int a = stack.top().pop();
				if(a < b){
					ip.left();
				}else if(a > b){
					ip.right();
				}
				break;
			}
			
			default:
				return super.execute(cmd);
		}
		return true;
	}
}

File: src/Befunge98Strategy.cpp
/**
 * @file Befunge98Strategy.cpp
 * Strategy for Befunge-98 commands.
 * @author Conlan Wesson
 */

#include "Befunge98Strategy.h"
#include "FungeRunner.h"

namespace Funge {


Befunge98Strategy::Befunge98Strategy(FungeRunner& r) :
	FungeStrategy(r)
{
	r.pushSemantic('[', std::bind(&Befunge98Strategy::instructionLeft, this), FungeSemantic::MOVEMENT);
	r.pushSemantic(']', std::bind(&Befunge98Strategy::instructionRight, this), FungeSemantic::MOVEMENT);
	r.pushSemantic('w', std::bind(&Befunge98Strategy::instructionCompare, this));
}

FungeError Befunge98Strategy::instructionLeft(){
	ip.left();
	if(runner.isMode(FUNGE_MODE_SWITCH)){
		ip.put(']');
	}
	return ERROR_NONE;
}

FungeError Befunge98Strategy::instructionRight(){
	ip.right();
	if(runner.isMode(FUNGE_MODE_SWITCH)){
		ip.put('[');
	}
	return ERROR_NONE;
}

FungeError Befunge98Strategy::instructionCompare(){
	stack_t b = stack.top().pop();
	stack_t a = stack.top().pop();
	if(a < b){
		ip.left();
	}else if(a > b){
		ip.right();
	}
	return ERROR_NONE;
}

FungeStrategy* Befunge98Strategy::clone(FungeRunner& r) const{
	return new Befunge98Strategy(r);
}

}
